cpp/src/shim/libcurl.cpp

```diff
@@ -116,19 +116,43 @@ CURL* CurlHandle::handle() noexcept { return _handle.get(); }
 
 void CurlHandle::perform()
 {
-  // Perform the curl operation and check for errors.
-  CURLcode err = curl_easy_perform(handle());
-  if (err != CURLE_OK) {
-    std::string msg(_errbuf);  // We can do this because we always initialize `_errbuf` as empty.
+  CURLcode err;
+  int http_code   = 0;
+  int retry_count = 0;
+  int max_retries = 5;
+
+  do {
     std::stringstream ss;
-    ss << "curl_easy_perform() error near " << _source_file << ":" << _source_line;
-    if (msg.empty()) {
-      ss << "(" << curl_easy_strerror(err) << ")";
+
+    CURLcode err = curl_easy_perform(handle());
+    curl_easy_getinfo(handle(), CURLINFO_RESPONSE_CODE, &http_code);
+
+    // Check if we should retry based on HTTP status code
+    if (http_code == 500 || http_code == 502 || http_code == 503 || http_code == 504) {
+      // Retry only if one of the specified status codes is returned
+      retry_count++;
+      ss << "Retrying (" << retry_count << "/" << max_retries << ") due to HTTP " << http_code
+         << std::endl;
+      if (retry_count >= max_retries) {
+        ss << "Max retries reached. Exiting." << std::endl;
+        throw std::runtime_error(ss.str());
+        break;
+      }
+    } else if (err != CURLE_OK) {
+      std::string msg(_errbuf);  // We can do this because we always initialize `_errbuf` as empty.
+      ss << "curl_easy_perform() error near " << _source_file << ":" << _source_line;
+      if (msg.empty()) {
+        ss << "(" << curl_easy_strerror(err) << ")";
+      } else {
+        ss << "(" << msg << ")";
+      }
+      throw std::runtime_error(ss.str());
     } else {
```

Comment:
This is unchanged from before, just moved after the HTTP status code check. Thinks like an overflow error in the callback handler might end up here and need to bubble up immediately.